src/Microsoft.OData.Core/ODataMessageReader.cs

```diff
@@ -887,29 +888,29 @@ internal ODataFormat GetFormat()
             return this.format;
         }
 
-        private static IServiceProvider GetContainer<T>(T message)
+        private static IServiceProvider GetServiceProvider<T>(T message)
             where T : class
         {
-            var containerProvider = message as IContainerProvider;
-            return containerProvider == null ? null : containerProvider.Container;
+            var serviceCollectionProvider = message as IServiceCollectionProvider;
+            return serviceCollectionProvider?.ServiceProvider;

```

Comment:
I don't know whether this is outside the scope of this PR, but I'm wondering why we retrieve the service provider from the `IODataRequest/ResponseMessage`. This seems to be the only reason why `IServiceCollectionProvider` interface exists. Is this the best way to approach this? Why not just pass the `IServiceProvider` to the `ODataMessageReader/Writer` constructor directly?

cc @ElizabethOkerio  @robertmclaws  @gathogojr

Reply:
I think this is outside the scope of this PR.

Reply:
While it may be outside the scope of this PR, would still like to get your thoughts on it. If you think the suggestion (to pass the container to the relevant constructors) is a better approach, then we could create a separate issue for it and possibly a PR for it in ODL 8 (or 9).

My issue with the current approach is that it's not intuitive at all. There's no way for a customer to expect (other than by reading the docs) that for them to pass custom settings to the `ODataMessageWriter`, they have to modify the `IODataRequest/ResponseMessage` to implement `IServiceCollectionProvider`. It also adds maintenance cost because it requires an extra public interface that in my opinion is unnecessary.

If passing the service provider to the constructor is not a favorable approach, the we could consider adding a property to the `ODataMessageReader/WriterSettings`

Reply:
This could be a design flaw and needs to be looked into. Let's create a work item for it.

Reply:
I think @habbes is onto something here but let's create a work item and then we can deliberate when to take it on for ODL 8 or after.

Reply:
In order to avoid derailing this PR, I've created a separate issue to track and discuss this: https://github.com/OData/odata.net/issues/2907